README
drbdmanage installation HowTo
==============================================================================


INSTALLATION
------------------------------------------------------------------------------

Prerequisites:
    DRBD 9 installed
    LVM installed
    python-dbus, python-gobject installed

IMPORTANT NOTICE:
The node names you enter must match the output of the "uname -n" command.

Installation procedure (all nodes):
    *  Install the drbdmanage package:
    *  Create a VG named drbdpool

Cluster initialization (first drbdmanage cluster node only):
    *  On the first node, issue the following command to initialize a new
       drbdmanage cluster:
         drbdmanage init <ipaddress>
       ... where <ipaddress> is the IP address of the network interface that
           DRBD should use to communicate with other nodes in the cluster

Adding nodes to a drbdmanage cluster:
    *  On a node of the existing drbdmanage cluster, issue the following
       command to prepare the cluster for integrating a new node:
         drbdmanage new-node <nodename> <ipaddress>
       ... where <nodename> is the node name of the node, and <ipaddress>
           is the IP address of the node that will join the cluster
       This command will also try to connect to the new node via ssh and
		 automatically join the cluster.
	 * If automatically joining is not possible: On the new node (the one that
		should join the cluster), enter the join command that you have
		received from the drbdmanage new-node command in the previous step (e.g., copy
		& paste)


CONFIGURATION
------------------------------------------------------------------------------

If you want to customize the settings of your drbdmanage server, you can
change its configuration by using 'drbdmanage modify-config'. Please refer to
drbdmanage(8) for further details.


Starting the drbdmanage server (daemon)
------------------------------------------------------------------------------

The drbdmanage server should be started automatically by D-Bus when the
drbdmanage client connects to it for the first time.

The following command can be used to check whether the drbdmanage client can
connect to the drbdmanage server, and it implicitly starts the server as
described above:
  drbdmanage ping

If the server has been started successfully or had been running already,
the response of the "ping" command is simply "pong"; otherwise, an
error message is displayed.

The server can also be started manually by issuing the following command:
  dbus-drbdmanaged-service &


drbdmanage commands overview
------------------------------------------------------------------------------

See manual page drbdmanage(8) for information on drbdmanage commands.

RELEASE/BUGFIX POLICY
------------------------------------------------------------------------------

Starting from drbdmanage 1.0.0 the project follows a three number version
scheme (x.y.z), where 'x' stands for "major", 'y' for "minor", and 'z' for a
bugfix release.

Releases that increase the major or minor number happen when development is
ready, we call them feature releases. Bugfix releases happen as soon as a bug
is discovered in a supported release, and increase the number 'z'.

- Support/Bugfixes:
We intend to support two consecutive releases. For example, as soon as we
release version 1.3.0, version 1.1.* is considered EoL and does not receive
further updates. This also applies when the major number is increased: Assume
the current version is 1.4.z, and then version 2.0.0 is released. In this case
we intend to support 2.0.* and 1.4.*. All versions prior to 1.4.* are EoL. We
support two releases overall, not two per major. If then 2.1.0 is released,
the whole 1.x.z release series is EoL.

- API/Data structure stability:
All supported versions with the same major number have a stable API. We might
add additional API calls, but we do not intentionally break/reduce the
existing API.

An increased major number means that we have introduced incompatible API
changes. We will document these changes in the release notes, but it is up to
3rd party developers to adapt their projects to the new API. Therefore, 'x' -
the major - can be considered as the API version.

- Upgrade path:
Upgrades are supported from one feature release to the next (e.g., from 1.2.*
to 1.3.*), as long as the major number stays the same.


debian/changelog
drbdmanage (0.99.18-1) experimental; urgency=medium

  * add before-resync-source, out-of-sync, quorum-lost handlers

 -- Roland Kammerer <roland.kammerer@linbit.com>  Wed, 04 Jul 2018 11:55:42 +0200

drbdmanage (0.99.17-1) experimental; urgency=medium

  [ Roland Kammerer ]
  * auto-deploy: fix case where resource ends in all diskless

 -- Roland Kammerer <roland.kammerer@linbit.com>  Mon, 07 May 2018 11:33:42 +0200

drbdmanage (0.99.16-1) experimental; urgency=medium

  [ Robert Altnoeder ]
  * Changed LvmThinLv space calculation

 -- Roland Kammerer <roland.kammerer@linbit.com>  Thu, 15 Mar 2018 11:28:21 +0100

drbdmanage (0.99.15-1) experimental; urgency=medium

  [ Lars Ellenberg ]
  * fix thin snapshot activation on some platforms: don't change the activ [...]

  [ Roland Kammerer ]
  * satellite: make timeouts configurable

 -- Roland Kammerer <roland.kammerer@linbit.com>  Wed, 14 Feb 2018 13:44:56 +0100

drbdmanage (0.99.14-1) experimental; urgency=medium

  [ Roland Kammerer ]
  * README -> README.md for github

  [ Robert Altnoeder ]
  * Fixed drbdadm resize command arguments order

 -- Roland Kammerer <roland.kammerer@linbit.com>  Fri, 13 Oct 2017 14:29:47 +0200

drbdmanage (0.99.13-1) experimental; urgency=medium

  [ Robert Altnoeder ]
  * Added free-port-nr and free-minor-nr commands
  * Added query-port-nr, query-minor-nr commands

  [ Roland Kammerer ]
  * client helper: add retry logic

 -- Roland Kammerer <roland.kammerer@linbit.com>  Tue, 26 Sep 2017 11:18:46 +0200

drbdmanage (0.99.12-1) experimental; urgency=medium

  * client: print "Waiting for" messages to stderr

 -- Roland Kammerer <roland.kammerer@linbit.com>  Thu, 21 Sep 2017 10:48:32 +0200

drbdmanage (0.99.11-1) experimental; urgency=medium

  [ Robert Altnoeder ]
  * Implemented separate fail counts for snapshot assignments

  [ Roland Kammerer ]
  * allow to keep ctrlvol on shutdown
  * leader: implement different error handling strategies
  * add more function to wrappers
    _____________________________
   / \                            \.
  |   |                           |.
   \_ |          I N G A          |.
      |             +             |.
      |         J A K O B         |.
      |                           |.
      |   ________________________|___
      |  /                           /.
      \_/___________________________/.
  * and happy birthday

 -- Roland Kammerer <roland.kammerer@linbit.com>  Fri, 15 Sep 2017 14:00:00 +0200

drbdmanage (0.99.10-1) experimental; urgency=medium

  * storage backends add notes about remove_snapshot()
  * provide zvol2/zvol_thinlv2 storage plugins
  * allow longer snapshot names

 -- Roland Kammerer <roland.kammerer@linbit.com>  Fri, 25 Aug 2017 10:24:58 +0200

drbdmanage (0.99.9-1) experimental; urgency=medium

  * fix mangling of dbus server return code

 -- Roland Kammerer <roland.kammerer@linbit.com>  Mon, 14 Aug 2017 11:12:28 +0200

drbdmanage (0.99.8-1) experimental; urgency=medium

  * fix locking between satellites/leader
  * allow actions on satellites only when triggered by leader
  * add wrapped return for list_assignments

 -- Roland Kammerer <roland.kammerer@linbit.com>  Wed, 09 Aug 2017 09:30:42 +0200

drbdmanage (0.99.7-1) experimental; urgency=medium

  * Automatic resource state correction before performing changes

 -- Roland Kammerer <roland.kammerer@linbit.com>  Tue, 01 Aug 2017 11:53:37 +0200

drbdmanage (0.99.6-1) experimental; urgency=medium

  [ Roland Kammerer ]
  * client helper: check for server startup
  * client helper: pep8/unused variable fixes
  * client helper: add its own exception type
  * fix too short timeout for leader/satellite communication

  [ Robert Altnoeder ]
  * Added debug commands for committing actions
  * Added resource/volume state debug output in _assignment_actions()
  * Added more debug output when deploying assignments
  * Modified debug commands for volumes, assignments and volume states
  * Added 'set bd'/'clear bd' debug commands
  * Enabled quorum tracking before control volume data is available

 -- Roland Kammerer <roland.kammerer@linbit.com>  Mon, 31 Jul 2017 13:07:38 +0200

drbdmanage (0.99.5-1) experimental; urgency=medium

  * client: fix bug in inline comments for ConfigParser
  * implement external nodes
  * wipefs: do not use "-q"
  * server: reset grace period on create_node
  * avoid shutting down satellites immediately
  * make cluster initialization more reliable

 -- Roland Kammerer <roland.kammerer@linbit.com>  Wed, 24 May 2017 14:02:10 +0200

drbdmanage (0.99.4-1) experimental; urgency=medium

  * introduce more fine grained error messages
  * wait-for/openstack: fix calls on satellite

 -- Roland Kammerer <roland.kammerer@linbit.com>  Mon, 06 Mar 2017 09:40:31 +0100

drbdmanage (0.99.3-1) experimental; urgency=medium

  [ Roland Kammerer ]
  * systemd: add dependency on network
  * fix path for wipefs
  * avoid creation of broken res files

  [ Robert Altnoeder ]
  * Always run drbdsetup down when undeploying resources

 -- Roland Kammerer <roland.kammerer@linbit.com>  Tue, 28 Feb 2017 14:46:02 +0100

drbdmanage (0.99.2-1) experimental; urgency=medium

  [ Roland Kammerer ]
  * LVM: use "wipefs" instead of "--wipesignatures=y"
  * fix startup on satellites

  [ Robert Altnoeder ]
  * Fall back to using existing volumes during deploy
  * Added debug flag for disabling the DrbdManager run() / perform_changes()

 -- Roland Kammerer <roland.kammerer@linbit.com>  Tue, 31 Jan 2017 13:56:49 +0100

drbdmanage (0.99.1-1) experimental; urgency=medium

  * leader (re)election: fix quorum tracking

 -- Roland Kammerer <roland.kammerer@linbit.com>  Mon, 23 Jan 2017 15:56:05 +0100

drbdmanage (0.99-1) experimental; urgency=medium

  [ Robert Altnoeder ]
  * Renamed variable 'signal' in some methods to avoid variable shadowing
  * Fix: restore_snapshot() decorator: @req_ctrlvol => @fwd_leader
  * Fix recycling minor numbers in get_free_minor_nr()

  [ Roland Kammerer ]

drbdmanage/consts.py
#!/usr/bin/env python2
"""
    drbdmanage - management of distributed DRBD9 resources
    Copyright (C) 2013 - 2017  LINBIT HA-Solutions GmbH
                               Author: R. Altnoeder, Roland Kammerer

    You can use this file under the terms of the GNU Lesser General
    Public License as as published by the Free Software Foundation,
    either version 3 of the License, or (at your option) any later
    version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU Lesser General Public License for more details.

    See <http://www.gnu.org/licenses/>.
"""

"""
Global constants for drbdmanage
"""

DM_VERSION = "0.99.18"
try:
    from drbdmanage.consts_githash import DM_GITHASH
except:
    DM_GITHASH = 'GIT-hash: UNKNOWN'

DBUS_DRBDMANAGED = "org.drbd.drbdmanaged"
DBUS_SERVICE     = "/interface"

DRBDADM_UTIL   = "drbdadm"
DRBDMETA_UTIL  = "drbdmeta"
DRBDSETUP_UTIL = "drbdsetup"

SERIAL              = "serial"
NODE_NAME           = "node_name"
NODE_ADDR           = "addr"
NODE_AF             = "addrfam"
NODE_ID             = "node_id"
NODE_VOL_0          = "drbdctrl_0"
NODE_VOL_1          = "drbdctrl_1"
NODE_ADDRESS        = "address"
NODE_PORT           = "drbdctrl_port"
NODE_SECRET         = "drbdctrl_secret"
NODE_SITE           = "site"
NODE_POOLSIZE       = "node_poolsize"
NODE_POOLFREE       = "node_poolfree"
NODE_STATE          = "node_state"
RES_NAME            = "res_name"
RES_PORT            = "port"
RES_SECRET          = "secret"
VOL_ID              = "vol_id"
VOL_MINOR           = "minor"
VOL_SIZE            = "vol_size"
VOL_BDEV            = "vol_bdev"
SNAPS_NAME          = "snaps_name"
ERROR_CODE          = "error_code"
FAIL_COUNT          = "fail-count"
COMMON_NAME         = "common_name"
MANAGED             = "managed"
CREATEDATE          = "create_date"

# Keys for the version text-query
KEY_SERVER_VERSION       = "server_version"
KEY_SERVER_GITHASH       = "server_git_hash"
KEY_DRBD_KERNEL_VERSION  = "drbd_kernel_version"
KEY_DRBD_KERNEL_GIT_HASH = "drbd_kernel_git_hash"
KEY_DRBD_UTILS_VERSION   = "drbd_utils_version"
KEY_DRBD_UTILS_GIT_HASH  = "drbd_utils_git_hash"

# Shut down resources on drbdmanage server shutdown
KEY_SHUTDOWN_RES = "shutdown-res"
# Drbdadm down ctrlvol
KEY_SHUTDOWN_CTRLVOL = "shutdown-ctrlvol"
# Error handling strategies
KEY_ERR_STRATEGY  = "error-handling-strategy"
KEY_ERR_RESUME_NO = "resume-no"
KEY_ERR_MAX_BOFF  = "resume-max-backoff"
KEY_ERR_INVTERVAL = "resume-interval"

SNAPS_SRC_BLOCKDEV  = "snapshot-source-blockdev"

# RFC952 / RFC1035 / RFC1123 host name constraints; do not change
NODE_NAME_MINLEN = 2
NODE_NAME_MAXLEN = 255
NODE_NAME_LABEL_MAXLEN = 63

# drbdmanage object name constraints
RES_NAME_MINLEN = 1
RES_NAME_MAXLEN = 48    # Enough for a UUID string plus prefix
RES_NAME_VALID_CHARS = "_"
RES_NAME_VALID_INNER_CHARS = "-"
RES_ALL_KEYWORD = "all"
SNAPS_NAME_MINLEN = 1
SNAPS_NAME_MAXLEN = 100
SNAPS_NAME_VALID_CHARS = "_"
SNAPS_NAME_VALID_INNER_CHARS = "-"

FAIL_COUNT_HARD_LIMIT = 99

DRBDCTRL_DEFAULT_PORT = 6999

KEY_DRBDCTRL_VG     = "drbdctrl-vg"
KEY_CUR_MINOR_NR    = "current-minor-nr"
KEY_VG_NAME         = "volume-group"
KEY_LOGLEVEL        = "loglevel"
DRBDCTRL_RES_NAME   = ".drbdctrl"
DRBDCTRL_RES_FILE   = "drbdctrl.res"
DRBDCTRL_DEV        = "/dev/drbd0"
DEFAULT_VG          = "drbdpool"
DRBDCTRL_RES_PATH   = "/etc/drbd.d/"
DRBDCTRL_DEV_0      = "/dev/drbd0"
DRBDCTRL_DEV_1      = "/dev/drbd1"
DRBDCTRL_LV_NAME_0  = ".drbdctrl_0"
DRBDCTRL_LV_NAME_1  = ".drbdctrl_1"

AF_IPV4 = 4
AF_IPV6 = 6

AF_IPV4_LABEL = "ipv4"
AF_IPV6_LABEL = "ipv6"

KEY_BLOCKSIZE = 'blocksize'
DEFAULT_BLOCKSIZE = '4k'

SERVER_CONFFILE     = "/etc/drbdmanaged.cfg"
KEY_DRBD_CONFPATH = "drbd-conf-path"
DEFAULT_DRBD_CONFPATH = "/var/lib/drbd.d"
FILE_GLOBAL_COMMON_CONF = "drbdmanage_global_common.conf"

# server instance
KEY_SERVER_INSTANCE = "serverinstance"

# additional configuration keys
KEY_SITE = 'site'

KEY_COLORS = "colors"
KEY_UTF8 = "utf8"

# auxiliary property prefix
AUX_PROP_PREFIX     = "aux:"

# flags prefixes
CSTATE_PREFIX       = "cstate:"
TSTATE_PREFIX       = "tstate:"

# resources, nodes, volumes:
FLAG_REMOVE         = "remove"

# nodes:
FLAG_UPD_POOL       = "upd_pool"
FLAG_UPDATE         = "update"
FLAG_DRBDCTRL       = "drbdctrl"
FLAG_STORAGE        = "storage"
FLAG_EXTERNAL       = "external"

# assignments, volume states:
FLAG_DEPLOY         = "deploy"

# assignments:
FLAG_DISKLESS       = "diskless"
FLAG_CONNECT        = "connect"
FLAG_UPD_CON        = "upd_con"
FLAG_RECONNECT      = "reconnect"
FLAG_OVERWRITE      = "overwrite"
FLAG_DISCARD        = "discard"
FLAG_UPD_CONFIG     = "upd_config"
FLAG_STANDBY        = "standby"
FLAG_QIGNORE        = "qignore"
FLAG_EXTERNAL       = "external"

IND_NODE_OFFLINE    = "node_offline"

# volume states:
FLAG_ATTACH         = "attach"

# reelection
FLAG_FORCEWIN = "forcewin"

# boolean expressions
BOOL_TRUE           = "true"
BOOL_FALSE          = "false"

RES_PORT_NR_AUTO    = -1
RES_PORT_NR_ERROR   = -2

TQ_GET_PATH         = "get_path"

CONF_GLOBAL = 'global'
CONF_NODE = 'node'
PLUGIN_PREFIX = 'Plugin:'

# ### satellites ###
FAKE_LEADER_NAME = '@@LEADER@@'  # just a fake name that is never a valid hostname
# is satellite, should be a control node but could not access ctrlvol, is control node
SAT_SATELLITE, SAT_POTENTIAL_LEADER_NODE, SAT_LEADER_NODE = range(3)
# cfg
KEY_SAT_CFG_SATELLITE = 'satellite'
KEY_SAT_CFG_CONTROL_NODE = 'controlnode'
KEY_SAT_CFG_ROLE = 'ctrl-volume-access-mode'
KEY_SAT_CFG_TCP_KEEPIDLE = 'tcp-keepidle'
KEY_SAT_CFG_TCP_KEEPINTVL = 'tcp-keepintvl'
KEY_SAT_CFG_TCP_KEEPCNT = 'tcp-keepcnt'
KEY_SAT_CFG_TCP_SHORTTIMEOUT = 'tcp-shorttimeout'
KEY_SAT_CFG_TCP_LONGTIMEOUT = 'tcp-longtimeout'

# after 10 sec of no other traffic,
# send a keep-alive every 7 seconds
# and fail if 5 not delivered
DEFAULT_SAT_CFG_TCP_KEEPIDLE = 10
DEFAULT_SAT_CFG_TCP_KEEPINTVL = 7
DEFAULT_SAT_CFG_TCP_KEEPCNT = 5
DEFAULT_SAT_CFG_TCP_SHORTTIMEOUT = 2.0
DEFAULT_SAT_CFG_TCP_LONGTIMEOUT = 45.0

# communication protocol
KEY_S_CMD_INIT = 'CMD_INIT'
KEY_S_CMD_UPDATE = 'CMD_UPDATE'
KEY_S_CMD_SHUTDOWN = 'CMD_SHUTDOWN'
KEY_S_CMD_PING = 'CMD_PING'
KEY_S_CMD_RELAY = 'CMD_RELAY'
KEY_S_CMD_REQCTRL = 'CMD_REQCTRL'
KEY_S_CMD_UPPOOL = 'CMD_UPPOOL'
KEY_S_INT_SHUTDOWN = 'INT_SHUTDOWN'
KEY_S_ANS_OK = 'ANS_OK'
KEY_S_ANS_E_LOCKING = 'ANS_LOCKING'
KEY_S_ANS_CHANGED = 'ANS_CHANGED'
KEY_S_ANS_CHANGED_FAILED = 'ANS_CHANGED_FAILED'
KEY_S_ANS_UNCHANGED = 'ANS_UNCHANGED'
KEY_S_ANS_E_OP_INVALID = 'ANS_OP_INVALID'
KEY_S_ANS_E_TOO_LONG = 'ANS_E_TOO_LONG'
KEY_S_ANS_E_COMM = 'ANS_E_COMM'

# BEGIN HOTFIX
# FIXME: Hotfix for resizing restored snapshots
#        after reducing the number of maximum peers
HOTFIX_MAX_PEERS = 31
# END HOTFIX
